src/utils/allocators.zig

```diff
@@ -2,6 +2,269 @@ const std = @import("std");
+                    record.is_free = true;
+                    return;
+                }
+            }
+            @panic("attempt to free invalid buf");
+        }
+
+        /// frees the unused space of a buf.
+        /// this is useful when a buf is initially overallocated and then resized.
+        pub fn tryRecycleUnusedSpace(self: *Self, buf_ptr: [*]T, used_len: u64) bool {
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            for (self.records.items) |*record| {
+                if (record.buf.ptr == buf_ptr) {
+                    return self.tryRecycleUnusedSpaceWithRecordUnsafe(record, used_len);
+                }
+            }
+            @panic("attempt to recycle invalid buf");
+        }
+
+        fn tryRecycleUnusedSpaceWithRecordUnsafe(self: *Self, record: *Record, used_len: u64) bool {
+            const unused_len = record.buf.len -| used_len;
+            if (unused_len > config.min_split_size) {
+                // update the state of the record
+                const split_buf = record.buf[used_len..];
+                // NOTE: this record ptr is updated before the append which could invalidate the record ptr
+                record.buf = record.buf[0..used_len];
+                record.len = used_len;
+                // add new unused record to the list
+                // NOTE: errors here are unreachable because if we hit OOM, were left in a bad state
+                self.records.append(.{
+                    .is_free = true,
+                    .buf = split_buf,
+                    .global_index = record.global_index + used_len,
+                    .len = split_buf.len,
+                    .memory_index = record.memory_index,
+                }) catch unreachable;
+                return true;
+            } else {
+                // dont try to split if its too small
+                return false;
+            }
+        }
+
+        /// collapses adjacent free records into a single record.
+        /// we use the memory_index to ensure we dont collapse two separate buffers
+        /// into one, which would result in a segfault.
+        pub fn collapseUnsafe(self: *Self) void {
+            const records = &self.records;
+            var i: usize = 1;
+            while (i < records.items.len) {
+                const prev = records.items[i - 1];
+                const curr = records.items[i];
+
+                const both_free = prev.is_free and curr.is_free;
+                const shared_memory_index = prev.memory_index == curr.memory_index;
+
+                if (both_free and shared_memory_index) {
+                    records.items[i - 1].buf.len += curr.buf.len;
+                    _ = records.orderedRemove(i);
+                } else {
+                    i += 1;
+                }
+            }
+        }
+
+        pub fn isPossibleToAllocateUnsafe(self: *Self, n: u64) bool {
```

Comment:
Similar to the other comment about the `Unsafe` suffix, not clear why this method is considered unsafe.

Reply:
Seems to be about thread safety. https://github.com/Syndica/sig/pull/381#discussion_r1846755166